### PR TITLE
Support generic stream/event on CUDA/HIP backend

### PR DESCRIPTION
**Pull request by guangyey**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #125757

# Motivation
According to [#123611](https://github.com/pytorch/pytorch/pull/123611), we support generic stream/event on CUDA backend.

# Additional Context
new method/attribute on `torch.Event` for cuda
- torch.Event.event_id
- torch.Event.elapsed_time
- torch.Event.synchronize

new method on `c10::Event` on cuda backend
- c10.Event.event_id
- c10.Event.elapsed_time
- c10.Event.synchronize

### CODE REVIEW

File: c10/cuda/impl/CUDAGuardImpl.h [hunk 1]
```diff
@@ -205,11 +205,35 @@ struct CUDAGuardImpl final : public c10::impl::DeviceGuardImplInterface {
...
     CUDACachingAllocator::recordStream(data_ptr, cuda_stream);
   }
+
+  double elapsedTime(void* event1, void* event2) const override {
+    TORCH_CHECK(
+        event1 && event2,
+        "Both events must be recorded before calculating elapsed time.");
+    cudaEvent_t cuda_event1 = static_cast<cudaEvent_t>(event1);
+    cudaEvent_t cuda_event2 = static_cast<cudaEvent_t>(event2);
+    float time_ms = 0;
```
Comment (albanD) [inline c10/cuda/impl/CUDAGuardImpl.h:232]: Could you add the device guard like https://github.com/pytorch/pytorch/blob/1ecea513b648502e4a0765256d165db5d3b42f52/aten/src/ATen/cuda/CUDAEvent.h#L157 here? Looks important.

**Review by albanD — APPROVED** [review]
Thanks!

**Review by jgong5 — CHANGES_REQUESTED** [review]
Test failing?

File: torch/csrc/Event.cpp [hunk 2]
```diff
@@ -58,8 +58,11 @@ static PyObject* THPEvent_pynew(
 
   new (&self->event) c10::Event(
       device->type(),
-      (enable_timing ? c10::EventFlag::PYTORCH_DEFAULT
-                     : c10::EventFlag::BACKEND_DEFAULT));
+      // See note [Flags defining the behavior of events]
```
Comment (guangyey) [inline torch/csrc/Event.cpp:61]: bug fix

File: aten/src/ATen/hip/impl/HIPGuardImplMasqueradingAsCUDA.h [hunk 3]
```diff
@@ -213,12 +213,35 @@ struct HIPGuardImplMasqueradingAsCUDA final : public c10::impl::DeviceGuardImplI
     hip_stream.synchronize();
   }
 
+  void synchronizeEvent(void* event) const override {
```
Comment (guangyey) [inline aten/src/ATen/hip/impl/HIPGuardImplMasqueradingAsCUDA.h:216]: help to support hip backend.

File: c10/cuda/impl/CUDAGuardImpl.h (hunk 1 above)
Comment (guangyey) [reply c10/cuda/impl/CUDAGuardImpl.h:241]: Updated. We also provide support on HIP backend.

**Review by jgong5 — APPROVED** [review]

**Review by EikanWang — APPROVED** [review]

File: c10/cuda/impl/CUDAGuardImpl.h (hunk 1 above)
Comment (albanD) [reply c10/cuda/impl/CUDAGuardImpl.h:241]: Why did you add get/set calls here and not `DeviceGuard guard(c10::kCUDA);` ?

File: aten/src/ATen/hip/impl/HIPGuardImplMasqueradingAsCUDA.h [hunk 4]
```diff
@@ -213,12 +213,36 @@ struct HIPGuardImplMasqueradingAsCUDA final : public c10::impl::DeviceGuardImplI
...
+  }
+
   void recordDataPtrOnStream(
     const c10::DataPtr& data_ptr,
     const Stream& stream) const override {
     HIPStreamMasqueradingAsCUDA hip_stream{stream};
     HIPCachingAllocatorMasqueradingAsCUDA::recordStreamMasqueradingAsCUDA(data_ptr, hip_stream);
   }
+
+  double elapsedTime(void* event1, void* event2, const DeviceIndex device_index)
```
Comment (albanD) [inline aten/src/ATen/hip/impl/HIPGuardImplMasqueradingAsCUDA.h:230]: Can we please make sure the relevant submodule owner has a chance to review the change before we merge it in other submodules?

cc @jeffdaily does this timing code looks good?

Comment (jeffdaily) [reply aten/src/ATen/hip/impl/HIPGuardImplMasqueradingAsCUDA.h:230]: Nothing jumps out as wrong.  I'll lean on CI to tell the truth :-).

File: c10/cuda/impl/CUDAGuardImpl.h (hunk 1 above)
Comment (guangyey) [reply c10/cuda/impl/CUDAGuardImpl.h:241]: In my opinion, `CUDAGuardImpl` is an implementation of `DeviceGuard`. `DeviceGuard` is high-level than `CUDAGuardImpl`. So it is unreasonable to use high-level code(`DeviceGuard`) in low-level code(`CUDAGuardImpl`).
If you prefer to use `DeviceGuard`, I can prepare a PR to refine it.

File: aten/src/ATen/hip/impl/HIPGuardImplMasqueradingAsCUDA.h (hunk 4 above)
Comment (guangyey) [reply aten/src/ATen/hip/impl/HIPGuardImplMasqueradingAsCUDA.h:230]: > Can we please make sure the relevant submodule owner has a chance to review the change before we merge it in other submodules?
> 
> cc @jeffdaily does this timing code looks good?

Sorry about that...

Comment (guangyey) [reply aten/src/ATen/hip/impl/HIPGuardImplMasqueradingAsCUDA.h:230]: I refer to `destroyEvent` code to guard it not to crate a new cuda context here. https://github.com/pytorch/pytorch/blob/b24ad7eab55eaf660893dddae949fc714e434338/c10/cuda/impl/CUDAGuardImpl.h#L106

**Pull request closed without merging**